datadog_lambda/tracing.py

```diff
@@ -235,37 +235,37 @@ def extract_context_from_sqs_or_sns_event_or_context(event, lambda_context):
 
         # logic to deal with SNS => SQS event
         if "body" in first_record:
-            body_str = first_record.get("body", {})
+            body_str = first_record.get("body")
             try:
                 body = json.loads(body_str)
                 if body.get("Type", "") == "Notification" and "TopicArn" in body:
                     logger.debug("Found SNS message inside SQS event")
                     first_record = get_first_record(create_sns_event(body))
             except Exception:
-                first_record = event.get("Records")[0]
                 pass
 
-        msg_attributes = first_record.get(
-            "messageAttributes",
-            first_record.get("Sns", {}).get("MessageAttributes", {}),
-        )
-        dd_payload = msg_attributes.get("_datadog", {})
+        msg_attributes = first_record.get("messageAttributes")
+        if msg_attributes is None:
+            sns_record = first_record.get("Sns") or {}
+            msg_attributes = sns_record.get("MessageAttributes") or {}
+        dd_payload = msg_attributes.get("_datadog")
         if dd_payload:
             # SQS uses dataType and binaryValue/stringValue
             # SNS uses Type and Value
             dd_json_data = None
-            dd_json_data_type = dd_payload.get("Type", dd_payload.get("dataType", ""))
+            dd_json_data_type = dd_payload.get("Type")
+            if dd_json_data_type is None:
+                dd_json_data_type = dd_payload.get("dataType")
             if dd_json_data_type == "Binary":
-                dd_json_data = dd_payload.get(
-                    "binaryValue",
-                    dd_payload.get("Value", r"{}"),
-                )
-                dd_json_data = base64.b64decode(dd_json_data)
+                dd_json_data = dd_payload.get("binaryValue")
+                if dd_json_data is None:
+                    dd_json_data = dd_payload.get("Value")
+                if dd_json_data:
+                    dd_json_data = base64.b64decode(dd_json_data)
             elif dd_json_data_type == "String":
-                dd_json_data = dd_payload.get(
-                    "stringValue",
-                    dd_payload.get("Value", r"{}"),
-                )
+                dd_json_data = dd_payload.get("stringValue")
+                if dd_json_data is None:
+                    dd_json_data = dd_payload.get("Value")
```

Comment:
Although I like that we are reducing allocations, this is really hard to read :cry: maybe in the future when we split the files, this would make more sense

Reply:
Yes agreed. This is difficult to read. I'll take a stab at improving it a bit.

Reply:
Another PR is fine, I guess when this is separated, this can be much easier! Not a blocking comment

Reply:
@duncanista Okay, I made some updates. Let me know if this is any better for you.